nautobot/nautobot_data_validation_engine/models.py

```diff
@@ -0,0 +1,355 @@
+"""Models for Data Validation Engine."""
+
+import re
+
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.core.validators import MinValueValidator, ValidationError
+from django.db import models
+from django.shortcuts import reverse
+
+try:
+    from nautobot.apps.constants import CHARFIELD_MAX_LENGTH
+except ImportError:
+    CHARFIELD_MAX_LENGTH = 255
+from nautobot.core.models.generics import PrimaryModel
+from nautobot.core.models.querysets import RestrictedQuerySet
+from nautobot.extras.utils import extras_features, FeatureQuery
+
+
+def validate_regex(value):
+    """
+    Checks that the value is a valid regular expression.
+
+    Don't confuse this with RegexValidator, which *uses* a regex to validate a value.
+    """
+    try:
+        re.compile(value)
+    except re.error as e:
+        raise ValidationError(f"{value} is not a valid regular expression.") from e
+
+
+class ValidationRuleManager(RestrictedQuerySet):
+    """Adds a helper method for getting all active instances for a given content type."""
+
+    def get_for_model(self, content_type):
+        """Given a content type string (<app_label>.<model>), return all instances that are enabled for that model."""
+        app_label, model = content_type.split(".")
+
+        return self.filter(content_type__app_label=app_label, content_type__model=model)
+
+
+class ValidationRule(PrimaryModel):
+    """Base model for all validation engine rule models."""
+
+    name = models.CharField(max_length=100, unique=True)
+    content_type = models.ForeignKey(
+        to=ContentType, on_delete=models.CASCADE, limit_choices_to=FeatureQuery("custom_validators")
+    )
+    enabled = models.BooleanField(default=True)
+    error_message = models.CharField(
+        max_length=255, blank=True, default="", help_text="Optional error message to display when validation fails."
+    )
+
+    objects = ValidationRuleManager.as_manager()
+
+    class Meta:
+        """Model metadata for all validation engine rule models."""
+
+        abstract = True
+
+    def __str__(self):
+        """Return a sane string representation of the instance."""
+        return self.name
+
+
+@extras_features(
+    "custom_fields",
+    "custom_links",
+    "export_templates",
+    "graphql",
+    "relationships",
+    "statuses",
+    "webhooks",
+)
+class RegularExpressionValidationRule(ValidationRule):
+    """A type of validation rule that applies a regular expression to a given model field."""
+
+    field = models.CharField(
+        max_length=50,
```

Comment:
Looks like the longest field name we have in core presently is `controller_device_redundancy_group` at 34 characters, but hopefully no apps are defining anything unreasonable.